WIP: 1710: Add SELinuxChangePolicy to PodSpec

jsafrane opened the pull request:
We found out that we *need* an explicit opt-in for mounting all volumes with `-o context` - all pods sharing the same volume must have the same label + must be all unprivileged (or all privileged).

There are many changes in the KEP because of that. It feels like a new KEP.

WIP:

* check PRR questionnaire + see what needs to be changed there
* split into smaller commits?

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/1710

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
Also - for this option to work, Pod must have an explicit SELinux context specified, otherwise it will again default to recursive. 

IMO - this default though makes sense, there isn't one single word that is going to capture all the necessary information. 

------

I was thinking - how about we keep existing default and introduce a boolean field in `securityContext`. So something like:

```securityContext.useSELInuxMountOption: true|false```

So we keep existing behaviour for `ReadWriteOncePod` and other situations (if field is `nil`) but use this field if explicitly set to true.  If this is explicitly set to `false`, then I guess we will have to use recursive chcon even for RWOP volumes. But that doesn't seem so bad.

jsafrane replied:
We should have all allowed options as explicit values in the API. `nil` should mean either `true` or `false`, not something in between.

I'll add the wording about SELinux label in Pod.

jsafrane replied:
Added a very short sentence about SELinuxLabel in PodSecurityContext / SecurityContext.

jsafrane commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -324,20 +381,20 @@ Therefore, if configured properly, all Pods in the same namespace run with the s
 2. They must run their CSI driver Pods with `/sys/fs/selinux` and `/etc/selinux/config` shared from the host via HostPath volumes!
    Because `/bin/mount` in the driver container evaluates these files and throws away any SELinux mount options if the files are not present.
 
 We will document this requirement in our documentation that faces CSI driver vendors in gihub.com/kubernetes-csi/docs.
```
Good point, replaced with a link.

jsafrane commented on keps/sig-storage/1710-selinux-relabeling/README.md:
Done

jsafrane commented on keps/sig-storage/1710-selinux-relabeling/README.md:
I;m not sure, "Volume Mounting" describes general mount capabilities. This example shows how we're going to use these capabilities in kubelet.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
Another option I was thinking is - `UseMountOptionsSafely`, it is perhaps as much big as current choice, but again this name is going to be just picking "okay" option among worst. :(

jsafrane replied:
Since the new default is opt-out, we don't need `UseMountOptionsSafely`

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
typo in `FSGroupChangePolicy` ?

jsafrane replied:
Fixed

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
IIRC - we wanted to clarify what happens if a privileged pod *does* specify selinux context explicitly. Does it get applied at all? Do we want to cover that case too?

gnufied replied:
ack. I missed it. I need to re-read everything again.

jsafrane commented:
Added details about missing kube-state-metrics + giant promql queries that would create a metric with CSI volume name and effective container SELinux label (with CSIDriver.spec.selinuxMount weighted in).

It should be possible to list volumes that are used with multiple SELinux labels.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
Nit: But we don't support selinux relabelling for hostpath volumes and hence pods that access hostpath this way must have elevated access already.

jsafrane replied:
The CVE is not about hostpath volumes. An attacker can create a bad symlink on say AWS EBS PV leading to `../../../..` and fool Kubernetes to provide it as subpath. The container runtime will then provide the files from the host and relabel them, because it thinks it's AWS EBS volume.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
I think if pod A and B, both use `MountOption` policy and they use different selinux contexts but get scheduled to different nodes, they both will be able to start. Previously - when using `Recursive` one pod would lock out the other.

jsafrane replied:
Good point, added a new "Multiple-node conflicts" section.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -74,27 +85,37 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
+2. All volumes are mounted with `-o context` by default, users can opt-out by setting `SELinuxChangePolicy: Recursive` in PodSpec.
+   With feature gate `SELinuxMount`, alpha without opt-out since 1.30, adding the opt-out field as alpha in 1.32.
+
+Initially, we thought we could do 2. without opt-out, but we found that it may break valid use cases.
```
nit: We are really implementing this in 3 phases.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -407,18 +568,24 @@ Due to change of Kubernetes behavior, we will implement the feature only for cas
+
+This phase went Beta (enabled by default) in Kubernetes 1.28, without `SELinuxChangePolicy` field in PodSpec!
+
+#### Phase 1.5
```
Phase 1.5 seems like a cop out. We implementing this really in 3 phases, there are no questions about it IMO.

jsafrane replied:
I added explicit phase 2 and 3 as a separate commit. I wanted the phases to be comparable with a previous version, but I agree that 1.5 is ugly.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -407,18 +568,24 @@ Due to change of Kubernetes behavior, we will implement the feature only for cas
+- Introduce `SELinuxChangePolicy` field, under `SELinuxChangePolicy` feature gate.
+  - In this phase, the only allowed value is `Recursive`.
+  - The field is not interpreted, it's used only by cluster admins to proactively opt-out from `SELinuxMount`
+  - We need to graduate it to Beta / enabled by default before enabling `SELinuxMount` feature gate.
```
So while `SELinuxChangePolicy` is enabled, what is the default in phase 2 (or phase 1.5)? I guess it defaults to `null` which has fluidity in its meaning. And are we planning to change the default in next release after this is enabled?

jsafrane replied:
`null` still means `MountOption`, but the whole field is not used by kubelet, only by metrics.

gnufied commented:
These changes look good to me and I know KEP is already pretty large but IMO few things that are missing:

1. Explicitly call out pods for which admins need to opt-out of `SELinuxMount` feature. I know there is a section that covers how conflicts between selinux contexts can occur(but that is more like mechanism of), but we need a user facing documentation which documents in simple words, if a pod will be affected.
2. I feel like we are missing a release plan.  So something like - v1.32 `SELinuxChangePolicy` goes alpha?, v1.33 `SELinuxChangePolicy` goes beta, `SELinuxMount` goes beta in v1.34 ?

jsafrane commented:
> I feel like we are missing a release plan. So something like - v1.32 SELinuxChangePolicy goes alpha?, v1.33 SELinuxChangePolicy goes beta, SELinuxMount goes beta in v1.34 ?

KEPs usually do not have a plan. They graduate as fast as they can, but the exact release nr. is not set anywhere.

jsafrane commented:
> Explicitly call out pods for which admins need to opt-out of SELinuxMount feature. I know there is a section that covers how conflicts between selinux contexts can occur(but that is more like mechanism of), but we need a user facing documentation which documents in simple words, if a pod will be affected. It is fine if we are going to cover this elsewhere.

I am interested too how should it be documented. A blog? Something that we can link from release notes.

gnufied commented:
/lgtm

gnufied commented:
> KEPs usually do not have a plan. They graduate as fast as they can, but the exact release nr. is not set anywhere.

Yes - but since we are going to be changing existing selinux behaviour and potentially breaking users, I thought it is best to be explicit about:

1. When we are going to deprecate existing behavior (perhaps we can deprecate it in v1.32 itself with clear announcement and release-notes).
2. When. we are going to ask users to opt-out (when `SELinuxChangePolicy` goes beta? (is this going to be beta but enabled by default or we are going to wait for GA of `SELinuxChangePolicy` feature-gate?)
3. When we are going to enable `SELinuxMount` as default feature and introduce new behaviour as the default? Will this be again disabled in beta and then enabled when GA?

This is going to take a long time I am afraid. We don't have to be specific about release numbers but it helps the reviewer(and people who need this feature) overall number of releases it will take to get to the end goal.


>I am interested too how should it be documented. A blog? Something that we can link from release notes.

Yeah - a blog sounds like a good idea. But perhaps we will need something more persistent on k/k website too.

deads2k commented on keps/sig-storage/1710-selinux-relabeling/README.md:
An event on pods that may get stuck is also a good idea to help workload authors.

jsafrane replied:
If we want events, we need a controller that sends them. kube-state-metrics used to emit metrics does not have any complex logic, all conflict detection would be done by promql queries.

deads2k commented on keps/sig-storage/1710-selinux-relabeling/README.md:
it'll be workload authors making the change, so both a metric and an event would help.

jsafrane replied:
Added a SELinuxController to kube-controller-manager to provide both events and metrics (so no kube-state-metrics changes are needed).

deads2k commented on keps/sig-storage/1710-selinux-relabeling/README.md:
Given that this will be the behavior otherwise, I'm not sure its necessary to have value. `nil` and `Recursive` seem to cover it.

deads2k commented:
Given scope limited to a subset of a subset (privileged and unprivileged mounting of the same PVC onSELinux clusters), the fact that we fail safely (pod stays in container creating), and the suggested mechanism gives a clear action signal to cluster-admins and workload authors both prior to the problem and gives a way to address the problem, this looks reasonable.

simonpasquier commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -588,21 +814,23 @@ _This section must be completed when targeting beta graduation to a release._
 
   - [x] Metrics
-    - All `errors_total` metrics below cover real errors when a Pod can't start.
+    - All `*_errors_total` metrics below cover real errors when a Pod can't start.
```
(suggestion) Can we explicit the labels for each metric?

Again it might be worth clarifying that all metrics are counters. I'm also not sure how the metrics would be used since they are not part of the rules below.

jsafrane replied:
Added labels + mentioned counters

simonpasquier commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -588,21 +814,23 @@ _This section must be completed when targeting beta graduation to a release._
-      1. `volume_manager_selinux_volume_context_mismatch_errors_total` + `volume_manager_selinux_volume_context_mismatch_warnings_total`: Number of errors when a Pod uses a volume that is already mounted with a different SELinux context than the Pod needs.
-         Before this feature, both pods would start, but only one such pod could access the volume.
-         With this feature, one of the Pods won't even start.
+        1. `volume_manager_selinux_container_errors_total` + `volume_manager_selinux_container_warnings_total`: Number of errors when kubelet cannot compute SELinux label for a container.
```
* The other metric names indicate the type of error ("..._pod_context_mismatch_...", "..._volume_context_mismatch_..."). Should we do the same here?
* Another option would be to have "volume_manager_selinux_errors_total"/"volume_manager_selinux_warnings_total" metric names and a specific `type=volume_context_mismatch|pod_context_mismatch` label but it depends if the rest of the label set is the same for all metrics.

jsafrane replied:
The labels are slightly different for each metric, depending if it is pod related (like `volume_manager_selinux_container_errors_total`) or volume related (then it has `volume_plugin` and `access_mode` label, so the admin can figure out if the issue is RWOP releated and which volume plugin / CSI driver has misconfigured `SELinuxMount` flag. I added list of labels to each metric here.

simonpasquier commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -818,26 +1054,166 @@ With a single `VolumeChangePolicy`, user has to fall back to `Recursive` policy
+      # The string is not very human friendly.
+      - record: kube_persistentvolume_info:unique_volume_id
+        expr: |
+          label_join(kube_persistentvolume_info, "unique_volume_id", "#", "gce_persistent_disk_name", "ebs_volume_id", "azure_disk_name", "fc_wwids", "fc_lun", "fc_target_wwns", "iscsi_target_portal", "iscsi_iqn", "iscsi_lun", "nfs_server", "nfs_path", "csi_driver", "csi_volume_handle", "local_path", "host_path")
```
it seems difficult to maintain? wouldn't it be hard to know all the label names in advance?

rexagod replied:
+1, this adds value only if the querying interface has a dropdown, in addition to the string itself being hard to read and write. The cardinality, which seems high and unbounded, stays the same if this is split into labels, which makes things convenient from the UX perspective (allowing the user to filter by `fc_wwids`, etc.).

rexagod replied:
I'd prefer having this as a metric in KSM where the value would be a hash instead, to counter both UX and cardinality issues.

jsafrane replied:
I moved the kube-state-metrics to "Alternatives" (a new controller in KCM with metrics + events is the new proposal).

simonpasquier commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -818,26 +1054,166 @@ With a single `VolumeChangePolicy`, user has to fall back to `Recursive` policy
+          kube_pod_security_context * on (pod, namespace, uid) group_right(pod_selinux_user, pod_selinux_role, pod_selinux_type, pod_selinux_level) kube_pod_container_security_context
+
+      # Compute correct SELinux label of a container in final_selinux_* labels and join it into final_selinux_label label that looks like "system_r:system_u:container_t:s0:c1,c92".
+      - record: kube_pod_container_security_context:final_selinux_label
```
I wonder if multi-stage recording rules wouldn't be more efficient/practical. E.g. something like (as long as we can ensure that the results for each recording rule don't overlap)

````
- record: kube_pod_container_security_context:final_selinux_label
  expr: |
    label_replace(
      kube_pod_container_security_context:pod_selinux{privileged="true"},
      "final_selinux_level",
      "s0",
      "",
      "")
- record: kube_pod_container_security_context:final_selinux_label
  expr: |
    label_replace(
      kube_pod_container_security_context:pod_selinux{container_selinux_level!="",privileged!="true"},
      "final_selinux_level",
      "$1",
      "container_selinux_level",
      "(.+)")
- ...
````

rexagod replied:
IIUC, and not weighing in any optimizations currently in place for this scenario by Thanos or other metrics storage solutions (since this is going into k/k), a metrics backend such as Prometheus, in and of itself, may end up [evaluating](https://github.com/prometheus/prometheus/blob/6e68046c257212e4259372dee4e97c3cc7cb0031/promql/engine.go#L1820) all split rules when the combined one is (a) queried or (b) hits its evaluation interval. I also want to point out that we may benefit from metrics that provide value on an individual level in at-least one scenario without necessarily needing to be joined, as such metrics may contribute to noise.

rexagod replied:
I believe this can be added under the pod collector in KSM.

rexagod replied:
I believe this can be added under the pod collector in KSM.

Edit: I just noticed that all rules are intermediaries for the last one. I believe we could add the first three rules (that don't require more than one core object collector) to KSM as IIUC despite being intermediaries, they seem to provide value on an individual level, while employing KSM's [CRS feature-set](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md) to [define](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/kube-state-metrics/custom-resource-state-configmap.yaml) the last rule (in OpenShift).

jsafrane replied:
I moved the kube-state-metrics to "Alternatives" (a new controller in KCM with metrics + events is the new proposal). I will revisit this if it gets back to the Proposal.
(Thanks for the hints!)

simonpasquier commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -346,59 +541,54 @@ To mitigate this risk, we propose:
+
+### kube-state-metrics
+
+*This section is under development. We need to figure out is Prometheus is able to do such complex queries.*
```
What would be the alternative if Prometheus can't handle it (which is going to be hard to determine)? Should the kubelet emit a gauge metric per container and/or pod instead?

rexagod commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -818,26 +1054,166 @@ With a single `VolumeChangePolicy`, user has to fall back to `Recursive` policy
+      # Add final_selinux_label to kube_pod_container_volume_mount.
+      # Augmented by csidriver.SELinuxMount, if set.
+      # This does a massive join from container -> pod -> pvc -> pv -> csidriver.
+      - record: kube_pod_container_volume_mount:selinux_label
+        expr: |
+          label_replace(
+            kube_pod_container_volume_mount
+              # Join final_selinux_label
+              * on(uid, pod, container) group_left (final_selinux_label) kube_pod_container_security_context:final_selinux_label
+              # Join CSIDriver.Spec.SELinuxMount by joining pvc -> pv -> csidriver
+              * on (uid, volume) group_left(persistentvolumeclaim) kube_pod_spec_volumes_persistentvolumeclaims_info
+              * on (namespace, persistentvolumeclaim) group_left(volumename) kube_persistentvolumeclaim_info
+              * on (volumename) group_left (csi_driver, csi_volume_handle, unique_volume_id) label_replace(kube_persistentvolume_info:unique_volume_id, "volumename", "$1", "persistentvolume", "(.+)")
+              * on(csi_driver) group_left (selinux_mount) kube_csidriver_info,
+                    
+            # label_replace: set final_selinux_label to "" when the CSI driver does not support SELinuxMount
+            "final_selinux_label", "", "selinux_mount", "false")
```
This is a bit more tricky since this depends on both PV and pod collectors, and as such, two different core objects' attributes, defining metrics for which at the same place is not supported in KSM by design.

jsafrane commented:
Added a new SELinuxController to kube-controller-manager to provide metrics and events. This controller should be opt-in forever, enabled by Kubernetes clusters that use SELinux. It is not required even on clusters with SELinux, it really just sends events and exposes metrics.

I moved kube-state-metrics to "Alternatives"

gnufied commented on keps/sig-storage/1710-selinux-relabeling/kep.yaml:
```diff
@@ -36,6 +37,9 @@ feature-gates:
       - kubelet
+  - name: SELinuxChangePolicy
+    components:
+      - kube-apiserver
```
nit: if we are adding a controller then, the flag should have to be enabled in KCM too right?

jsafrane replied:
good catch, added

gnufied commented:
please fix the places where feature-gate needs to be enabled. Other than that lgtm.

/lgtm

deads2k commented:
Thanks for the metrics and events.  PRR looks good.

/approve

jsafrane commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -311,33 +454,85 @@ spec:
+  * It is emitted only when the `SELinuxMount` feature gate is enabled.
+  * It is emitted by kubelet when a Pod really cannot start.
+  * It misses Pods that run on different nodes that would not run if they landed on the same node.
+  * Since kubelet re-tries to start Pods periodically, this metric will grow until the Pod starts or is deleted.
```
Fixed

jsafrane commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -311,33 +454,85 @@ spec:
-* Implement this feature in [several phases](#Implementation-phases).
-* Expose metrics + alerts before the feature is enabled by default and could break anyone.
+Phase 3:
+* `volume_manager_selinux_volume_context_mismatch_errors_total`: nr. of Pods that did not start.
```
Fixed.

dipesh-rawat commented on keps/sig-storage/1710-selinux-relabeling/kep.yaml:
```diff
@@ -26,6 +26,7 @@ milestone:
   stable: "v1.31" # Very optimistic plan for SELinuxMountReadWriteOncePod GA
 
   # alpha: "v1.30"  # SELinuxMount
+  # alpha: "v1.32"  # SELinuxChangePolicy
```
Just a small question - should the latest milestone be updated to `latest-milestone: "v1.32"` to align with this update?

Below is the reference from the KEP [template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/kep.yaml), so wanted to clarify. Apologies for any confusion on my part!

https://github.com/kubernetes/enhancements/blob/3e50abe0a97ea57f79e05557ac316421deceb800/keps/NNNN-kep-template/kep.yaml#L27-L30

jsafrane replied:
good point, fixed

gnufied commented:
/lgtm
